### PR TITLE
LEB128 and Function/Body Split

### PR DESCRIPTION
**Pull request by commonlisp**
- Added decoding of signed LEB128
- Using LEB128 for counts, indices, and offsets in several places
- Splitting up function declaration (function header)
  and function body

### CODE REVIEW

**Comment by MikeHolman** [comment]
lgtm

File: lib/WasmReader/WasmBinaryReader.cpp [hunk 1]
```diff
@@ -137,20 +137,28 @@ WasmBinaryReader::ReadFromModule()
...
             m_moduleState.count++;
+            break;
 
+        case bSectFunBodies:
+            if (!m_visitedSections->Test(bSectFunSigs))
+            {
+                ThrowDecodingError(L"Function signatures section missing before function bodies");
+            }
+            FunctionBodyHeader();
+            m_moduleState.count++;
```
Comment (nmostafa) [inline lib/WasmReader/WasmBinaryReader.cpp]: Shouldn't we check that the count of FuncSigs and FuncBodies match ?

File: lib/WasmReader/WasmBinaryReader.cpp [hunk 2]
```diff
@@ -530,39 +546,63 @@ WasmBinaryReader::FunctionHeader()
...
+            break;
+        default:
+            ThrowDecodingError(L"Unexpected local type");
         }
     }
 
+    for (UINT i = 0; i < i32Count; i++)
+    {
+        m_funcInfo->AddLocal(Wasm::WasmTypes::I32);
+    }
```
Comment (nmostafa) [inline lib/WasmReader/WasmBinaryReader.cpp]: Maybe I am missing something, but shouldn't' we have a list of funcInfos signatures contructed from FuncSigs that we attribute function bodies to here ? 
See [Luke's PR](https://github.com/WebAssembly/design/blob/0d96ed985368fc636a38de31970dc35274018ada/BinaryEncoding.md#function-bodies-section)

Comment (MikeHolman) [inline lib/WasmReader/WasmBinaryReader.cpp]: locals are defined in the body, not the signature

Comment (nmostafa) [inline lib/WasmReader/WasmBinaryReader.cpp]: Not what I meant. There are two separate sections now: One for signatures, and another for bodies. They are associated by the order of the entries. So if the function signature section has 3 functions, the reader should internally keep a list of 3 functions, then go back an add locals to them. Currently, I see one funcInfo tracked by the reader.

Comment (MikeHolman) [inline lib/WasmReader/WasmBinaryReader.cpp]: oh, you're absolutely right. We should do `moduleInfo->GetFuncInfo(n)` lookup.

File: lib/WasmReader/WasmBinaryReader.cpp (hunk 1 above)
Comment (commonlisp) [inline lib/WasmReader/WasmBinaryReader.cpp]: Good catch. I'm checking that now.

**Pull request merged**